{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1eac4e3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2026-01-14 14:00:01--  https://ftp.ensembl.org/pub/release-114/fasta/mus_musculus/cdna/Mus_musculus.GRCm39.cdna.all.fa.gz\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 50893631 (49M) [application/x-gzip]\n",
      "Saving to: ‘Mus_musculus.GRCm39.cdna.all.fa.gz’\n",
      "\n",
      "Mus_musculus.GRCm39 100%[===================>]  48.54M  11.3MB/s    in 4.8s    \n",
      "\n",
      "2026-01-14 14:00:06 (10.0 MB/s) - ‘Mus_musculus.GRCm39.cdna.all.fa.gz’ saved [50893631/50893631]\n",
      "\n",
      "--2026-01-14 14:00:07--  https://ftp.ensembl.org/pub/release-114/fasta/mus_musculus/ncrna/Mus_musculus.GRCm39.ncrna.fa.gz\n",
      "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.169\n",
      "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.169|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 28414607 (27M) [application/x-gzip]\n",
      "Saving to: ‘Mus_musculus.GRCm39.ncrna.fa.gz’\n",
      "\n",
      "Mus_musculus.GRCm39 100%[===================>]  27.10M  7.38MB/s    in 3.7s    \n",
      "\n",
      "2026-01-14 14:00:11 (7.38 MB/s) - ‘Mus_musculus.GRCm39.ncrna.fa.gz’ saved [28414607/28414607]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget https://ftp.ensembl.org/pub/release-114/fasta/mus_musculus/cdna/Mus_musculus.GRCm39.cdna.all.fa.gz\n",
    "!gunzip Mus_musculus.GRCm39.cdna.all.fa.gz\n",
    "!wget https://ftp.ensembl.org/pub/release-114/fasta/mus_musculus/ncrna/Mus_musculus.GRCm39.ncrna.fa.gz\n",
    "!gunzip Mus_musculus.GRCm39.ncrna.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3c70e5ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "from baitshop import code_design, barcode_assignment\n",
    "import pandas as pd\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "import numpy as np\n",
    "import Bio.SeqIO as SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "556632da",
   "metadata": {},
   "outputs": [],
   "source": [
    "average_expression = pd.read_csv('data/Mean_expression_per_cluster.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "957f0d87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Identify genes with >4000 counts in any cluster (genes are indices)\n",
    "high_expr_genes = average_expression[average_expression.max(axis=1) > 4000].index.tolist()\n",
    "high_expr_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2b815748",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved 139 codewords to ljcr_139_hd4.csv\n"
     ]
    }
   ],
   "source": [
    "# Download a 16 bit LJCR code with k=4 and t=3\n",
    "code_design.download_ljcr_csv(v=16, k=4, t=3, output_path='ljcr_139_hd4.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2805228b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter codewords with Hamming distance < 4\n",
    "codewords = code_design.filter_hd_gt4('ljcr_139_hd4.csv')\n",
    "len(codewords)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dc2cc6d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "codeword_array = np.array([list(map(int, bc)) for bc in codewords])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2fb2a17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "codeword_list = [list(map(int, bc)) for bc in codewords]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3c3ec7aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "139"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create codeword_list as list of strings\n",
    "codeword_strings = [''.join(map(str, bc)) for bc in codeword_array]\n",
    "len(codeword_strings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "397de9a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0., 4., 4., ..., 8., 8., 8.],\n",
       "       [4., 0., 4., ..., 8., 8., 8.],\n",
       "       [4., 4., 0., ..., 8., 8., 8.],\n",
       "       ...,\n",
       "       [8., 8., 8., ..., 0., 4., 4.],\n",
       "       [8., 8., 8., ..., 4., 0., 4.],\n",
       "       [8., 8., 8., ..., 4., 4., 0.]])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get hamming distances between codewords\n",
    "hamming_condensed = pdist(codeword_array, metric='hamming')\n",
    "hamming_matrix = squareform(hamming_condensed) * codeword_array.shape[1]\n",
    "hamming_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b326638f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get bulk RNA-seq isoform expression data, APPRIS data, and gene names\n",
    "isoform_expression = pd.read_csv('data/Zhang_2023_Salmon_TPM.csv')\n",
    "appris_data = pd.read_table('data/appris_data.principal.txt', sep='\\t', header=None)\n",
    "with open('data/gene_names.txt') as f:\n",
    "    gene_names = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "85f3d815",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename columns for clarity\n",
    "isoform_expression.columns = ['Name', 'TPM']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "8aa54e88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>isoform</th>\n",
       "      <th>annotation_level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sf3b1</td>\n",
       "      <td>ENSMUSG00000025982</td>\n",
       "      <td>ENSMUST00000027127</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fev</td>\n",
       "      <td>ENSMUSG00000055197</td>\n",
       "      <td>ENSMUST00000068631</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Septin2</td>\n",
       "      <td>ENSMUSG00000026276</td>\n",
       "      <td>ENSMUST00000027495</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Atf6</td>\n",
       "      <td>ENSMUSG00000026663</td>\n",
       "      <td>ENSMUST00000027974</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nck2</td>\n",
       "      <td>ENSMUSG00000066877</td>\n",
       "      <td>ENSMUST00000086421</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33950</th>\n",
       "      <td>mt-Nd4l</td>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>ENSMUST00000084013</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33951</th>\n",
       "      <td>mt-Nd5</td>\n",
       "      <td>ENSMUSG00000064367</td>\n",
       "      <td>ENSMUST00000082418</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33952</th>\n",
       "      <td>mt-Nd4</td>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>ENSMUST00000082414</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33953</th>\n",
       "      <td>mt-Nd6</td>\n",
       "      <td>ENSMUSG00000064368</td>\n",
       "      <td>ENSMUST00000082419</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33954</th>\n",
       "      <td>mt-Cytb</td>\n",
       "      <td>ENSMUSG00000064370</td>\n",
       "      <td>ENSMUST00000082421</td>\n",
       "      <td>PRINCIPAL:1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33955 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene             gene_id             isoform annotation_level\n",
       "0        Sf3b1  ENSMUSG00000025982  ENSMUST00000027127      PRINCIPAL:1\n",
       "1          Fev  ENSMUSG00000055197  ENSMUST00000068631      PRINCIPAL:1\n",
       "2      Septin2  ENSMUSG00000026276  ENSMUST00000027495      PRINCIPAL:1\n",
       "3         Atf6  ENSMUSG00000026663  ENSMUST00000027974      PRINCIPAL:1\n",
       "4         Nck2  ENSMUSG00000066877  ENSMUST00000086421      PRINCIPAL:1\n",
       "...        ...                 ...                 ...              ...\n",
       "33950  mt-Nd4l  ENSMUSG00000065947  ENSMUST00000084013      PRINCIPAL:1\n",
       "33951   mt-Nd5  ENSMUSG00000064367  ENSMUST00000082418      PRINCIPAL:1\n",
       "33952   mt-Nd4  ENSMUSG00000064363  ENSMUST00000082414      PRINCIPAL:1\n",
       "33953   mt-Nd6  ENSMUSG00000064368  ENSMUST00000082419      PRINCIPAL:1\n",
       "33954  mt-Cytb  ENSMUSG00000064370  ENSMUST00000082421      PRINCIPAL:1\n",
       "\n",
       "[33955 rows x 4 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Rename columns and remove '3'\n",
    "appris_data.drop(columns=[3], inplace=True)\n",
    "appris_data.columns = ['gene','gene_id', 'isoform', 'annotation_level']\n",
    "appris_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6d3b7c17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>isoform</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Igkv1-133</td>\n",
       "      <td>ENSMUST00000103304.3</td>\n",
       "      <td>374</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Igkv1-132</td>\n",
       "      <td>ENSMUST00000103305.2</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Igkv1-131</td>\n",
       "      <td>ENSMUST00000103306.2</td>\n",
       "      <td>362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Igkv17-127</td>\n",
       "      <td>ENSMUST00000200586.2</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Igkv17-127</td>\n",
       "      <td>ENSMUST00000103309.3</td>\n",
       "      <td>341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116308</th>\n",
       "      <td>Gm14415</td>\n",
       "      <td>ENSMUST00000118260.2</td>\n",
       "      <td>1578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116309</th>\n",
       "      <td>Gm14404</td>\n",
       "      <td>ENSMUST00000119117.2</td>\n",
       "      <td>1660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116310</th>\n",
       "      <td>Gm16357</td>\n",
       "      <td>ENSMUST00000144091.2</td>\n",
       "      <td>474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116311</th>\n",
       "      <td>Gm13339</td>\n",
       "      <td>ENSMUST00000120651.2</td>\n",
       "      <td>510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116312</th>\n",
       "      <td>Spcs2-ps</td>\n",
       "      <td>ENSMUST00000122326.2</td>\n",
       "      <td>682</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>116313 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              gene               isoform  length\n",
       "0        Igkv1-133  ENSMUST00000103304.3     374\n",
       "1        Igkv1-132  ENSMUST00000103305.2     362\n",
       "2        Igkv1-131  ENSMUST00000103306.2     362\n",
       "3       Igkv17-127  ENSMUST00000200586.2     359\n",
       "4       Igkv17-127  ENSMUST00000103309.3     341\n",
       "...            ...                   ...     ...\n",
       "116308     Gm14415  ENSMUST00000118260.2    1578\n",
       "116309     Gm14404  ENSMUST00000119117.2    1660\n",
       "116310     Gm16357  ENSMUST00000144091.2     474\n",
       "116311     Gm13339  ENSMUST00000120651.2     510\n",
       "116312    Spcs2-ps  ENSMUST00000122326.2     682\n",
       "\n",
       "[116313 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get transcript lengths from cDNA FASTA file\n",
    "# Isoform is record.id, gene is record.description gene_symbol:\"gene\"\n",
    "data = []\n",
    "for record in SeqIO.parse('Mus_musculus.GRCm39.cdna.all.fa', 'fasta'):\n",
    "\tisoform = record.id.split('|')[0]\n",
    "\tlength = len(record.seq)\n",
    "\t# Extract gene symbol from description\n",
    "\tif 'gene_symbol:' in record.description:\n",
    "\t\tgene = record.description.split('gene_symbol:')[1].split()[0]\n",
    "\telse:\n",
    "\t\tgene = None\n",
    "\tdata.append({'gene': gene, 'isoform': isoform, 'length': length})\n",
    "\n",
    "transcript_lengths = pd.DataFrame(data)\n",
    "transcript_lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "93391abb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make sure every gene in gene_names is in transcript_lengths\n",
    "missing_genes = [gene for gene in gene_names if gene not in transcript_lengths['gene'].values]\n",
    "missing_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "3014135a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the best isoform for each gene\n",
    "best_isoforms = code_design.select_best_isoform(isoform_expression, gene_names, appris_data, transcript_lengths)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f5d2993f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load Spearman correlation matrix\n",
    "spearman_mat = np.load('data/Gene_Spearman.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "697fee8e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(spearman_mat[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2ec2ac98",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign barcodes using low-rank Gromov-Wasserstein method\n",
    "barcode_assignments = barcode_assignment.assign_barcodes_lowrank_gw(genes=gene_names, barcodes=codeword_strings,\n",
    "                                                                    correlation_matrix=spearman_mat, distance_matrix=hamming_matrix,epsilon=1e-05,\n",
    "                                                                    max_iterations=1000, variance_fraction=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "675125d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'mean_distance_high_corr': 6.357142857142857,\n",
       " 'mean_distance_med_corr': 5.972754793138244,\n",
       " 'mean_distance_low_corr': 5.987074829931973,\n",
       " 'min_distance_high_corr': 4.0,\n",
       " 'correlation_distance_pearson': 0.009832522845345677,\n",
       " 'correlation_distance_spearman': 0.003816516813200746}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Evaluate the barcode assignment\n",
    "eval = barcode_assignment.evaluate_assignment(correlation_matrix=spearman_mat,\n",
    "                                             distance_matrix=hamming_matrix,\n",
    "                                             assignment=barcode_assignments)\n",
    "eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "523cd7bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create dict of readout names to sequences\n",
    "readouts = SeqIO.parse('data/69_bit_readouts.fasta', 'fasta')\n",
    "readout_dict = {record.id: str(record.seq) for record in readouts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "268310d7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['RS0015', 'RS0083', 'RS0095', 'RS0109', 'RS0175', 'RS0237', 'RS0247', 'RS0255', 'RS0307', 'RS0332', 'RS0343', 'RS0384', 'RS0406', 'RS0451', 'RS0468', 'RS0548', 'RS0578', 'RS0584', 'RS0639', 'RS0707', 'RS0708', 'RS0730', 'RS0763', 'RS0793', 'RS0805', 'RS0896', 'RS0820', 'RS0994', 'RS0996', 'RS0967', 'RS1040', 'RS1047', 'RS1199', 'RS1316', 'RS1312', 'RS1261', 'RS1326', 'RS1404', 'RS1334', 'RS3505', 'RS1470', 'RS1520', 'RS3583', 'RS1678', 'RS3592', 'RS3602', 'RS1982', 'RS3791', 'RS3811', 'RS3918', 'RS3544', 'RS4013', 'RS4027', 'RS4070', 'RS4098', 'RS4111', 'RS4140', 'RS4164', 'RS4194', 'RS4201', 'RS4216', 'RS4324', 'RS4341', 'RS4351', 'RS4363', 'RS4365', 'RS4465', 'RS4483', 'RS4554'])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "readout_dict.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "741ac038",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create the codebook\n",
    "codebook = code_design.create_codebook(genes=gene_names, barcodes=codeword_strings, assignment=barcode_assignments,\n",
    "                                       isoform_df=best_isoforms, readouts=readout_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "57f551d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>id</th>\n",
       "      <th>RS0015</th>\n",
       "      <th>RS0083</th>\n",
       "      <th>RS0095</th>\n",
       "      <th>RS0109</th>\n",
       "      <th>RS0175</th>\n",
       "      <th>RS0237</th>\n",
       "      <th>RS0247</th>\n",
       "      <th>RS0255</th>\n",
       "      <th>RS0307</th>\n",
       "      <th>RS0332</th>\n",
       "      <th>RS0343</th>\n",
       "      <th>RS0384</th>\n",
       "      <th>RS0406</th>\n",
       "      <th>RS0451</th>\n",
       "      <th>RS0468</th>\n",
       "      <th>RS0548</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adarb2</td>\n",
       "      <td>ENSMUST00000135574.8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Erbb4</td>\n",
       "      <td>ENSMUST00000121473.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Trpm3</td>\n",
       "      <td>ENSMUST00000237357.2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Meis2</td>\n",
       "      <td>ENSMUST00000074285.8</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nrxn3</td>\n",
       "      <td>ENSMUST00000163134.8</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>Grm5</td>\n",
       "      <td>ENSMUST00000125009.9</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Gria1</td>\n",
       "      <td>ENSMUST00000036315.16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Arpp21</td>\n",
       "      <td>ENSMUST00000161467.8</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Chrm3</td>\n",
       "      <td>ENSMUST00000187510.7</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Sox5</td>\n",
       "      <td>ENSMUST00000038815.14</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      name                     id  RS0015  RS0083  RS0095  RS0109  RS0175  \\\n",
       "0   Adarb2   ENSMUST00000135574.8       0       1       0       0       0   \n",
       "1    Erbb4   ENSMUST00000121473.8       0       0       0       0       0   \n",
       "2    Trpm3   ENSMUST00000237357.2       0       0       1       1       1   \n",
       "3    Meis2   ENSMUST00000074285.8       1       1       0       0       0   \n",
       "4    Nrxn3   ENSMUST00000163134.8       0       1       0       0       0   \n",
       "..     ...                    ...     ...     ...     ...     ...     ...   \n",
       "95    Grm5   ENSMUST00000125009.9       0       0       1       1       0   \n",
       "96   Gria1  ENSMUST00000036315.16       0       0       1       0       1   \n",
       "97  Arpp21   ENSMUST00000161467.8       0       0       0       0       0   \n",
       "98   Chrm3   ENSMUST00000187510.7       0       0       1       1       0   \n",
       "99    Sox5  ENSMUST00000038815.14       0       1       0       0       0   \n",
       "\n",
       "    RS0237  RS0247  RS0255  RS0307  RS0332  RS0343  RS0384  RS0406  RS0451  \\\n",
       "0        0       1       0       1       0       0       0       0       0   \n",
       "1        1       1       0       0       0       0       0       0       1   \n",
       "2        1       0       0       0       0       0       0       0       0   \n",
       "3        0       0       0       0       0       0       0       1       1   \n",
       "4        0       1       0       1       0       0       0       0       0   \n",
       "..     ...     ...     ...     ...     ...     ...     ...     ...     ...   \n",
       "95       0       0       0       0       0       1       1       0       0   \n",
       "96       0       0       0       0       0       0       1       0       1   \n",
       "97       0       0       0       1       1       0       0       1       1   \n",
       "98       0       0       0       0       0       0       0       0       0   \n",
       "99       0       0       1       0       0       0       1       0       1   \n",
       "\n",
       "    RS0468  RS0548  \n",
       "0        0       1  \n",
       "1        1       0  \n",
       "2        0       0  \n",
       "3        0       0  \n",
       "4        0       1  \n",
       "..     ...     ...  \n",
       "95       0       0  \n",
       "96       0       0  \n",
       "97       0       0  \n",
       "98       1       1  \n",
       "99       0       0  \n",
       "\n",
       "[100 rows x 18 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codebook"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "MERFISH_probe_design",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
